File: completed-labs/18/java/Trie.java
public class Trie {

	public static final int alphabet_size = 26;

    public Trie[] children;
    boolean endWord;
    public Trie() {
    	this.children = new Trie[alphabet_size+1];
    	this.endWord = false;
    	for(int i=0;i<children.length;i++){
    		children[i] = null;
    	}
    }
    public void insert(String s) {
    	int curIndex;

    	//current trie
    	Trie crawl = this;

    	//loop through input string and add to current's array
    	for(int i = 0;i<s.length();i++){
    		curIndex = s.charAt(i) -'a';
    		if(crawl.children[curIndex] == null){
    			crawl.children[curIndex] = new Trie();
    		}
    		crawl = crawl.children[curIndex];
    	}
    		curIndex = '{' - 'a';

    		crawl.children[curIndex] = new Trie();
    		crawl.endWord = true;
    }
    public boolean query(String s) {
    	Trie temp = this;

    	for(int i=0; i<s.length();i++){
    		int index = s.charAt(i) - 'a';
    		temp = temp.children[index];
    		if(temp == null){
    			return false;
    		}
    	}

	return (temp.endWord && temp != null);
    }
}

File: completed-labs/18/cpp/Trie.cpp
#include "Trie.hpp"

Trie::Trie() {
    this -> isLeaf = false;
    this -> children = new Trie* [27]; 
    for (int i = 0; i < 27; i++) {
        this -> children[i] = NULL; 
    }
}

void Trie::insert(std::string s) {        
    Trie* temp = this;
    int index;
    for (int i = 0; i < s.length(); i++) {
        index = s[i] - 'a';

        if (temp -> children[index] == NULL){
            temp -> children[index] = new Trie();
        }
        temp = temp -> children[index];
    }
    index = '{' -'a';
    temp -> children[index] = new Trie();
    
    temp -> isLeaf = true;
}

bool Trie::query(std::string s) {
    Trie* temp = this;
    for (int i = 0; i < s.length(); i++) {
        int index = s[i] - 'a';
        temp = temp -> children[index];

        if (temp == NULL) {
            return false;
        }
    }
    return temp -> isLeaf;
}
